src/target/riscv/riscv-013.c

```diff
@@ -1712,8 +1741,17 @@ static int examine(struct target *target)
 	}
 
 	riscv013_info_t *info = get_info(target);
-	/* TODO: This won't be true if there are multiple DMs. */
-	info->index = target->coreid;
+	/* As info->index is the index of each hart on the DM (Debug Module),
+	 * for each DM these must start at 0 and be contiguous.
+	 * And target->coreid is target->coreid is globally unique in OpenOCD.
+	 * So for multiple DMs, info->index and target->coreid can't be the same.
+	 * Consider different DM have different dbgbase, let coreid = index + dbgbase */
+	if ((unsigned int)target->coreid < target->dbgbase) {
+		LOG_TARGET_ERROR(target, "coreid should not be smaller than DMI base address (dbgbase), coreid=0x%x dbgbase=0x%x",
+				target->coreid, target->dbgbase);
+		return ERROR_FAIL;
+	}
+	info->index = target->coreid - target->dbgbase;
```

Comment:
This does not seem right to me. I believe coreids and debug module base addresses should be independent.


For example, consider debug module which controls 1024 harts and base DMI address 0x0. Base DMI address of the next debug module could be for example 0x100. It does not have to be 0x400 (which would be 1024 in decimal).

Reply:
Yes, coreid and dbgbase really don't have to be associated, I just want to find a simple way to map unique coreid to index. But as you said, this mapping method is not appropriate for some case, I will think about it again. 
Thank you for pointing it out.

Reply:
Up to `2^20` harts can be connected to a single DM, how about use higher bits to distinguish between different DMs?

That means `info->index = target->coreid & (1<<20 - 1)`

Reply:
As long as the high bits are a DM index and not a DM address then that storage scheme seems fine.

It's a bit user-unfriendly to have to write:
`target create $_TARGETNAME_1 riscv -chain-position $_CHIPNAME.cpu -coreid 1048577`

But we could address that separately by introducing a new tcl command that takes separate DM index and debug index. It won't look pretty either, but again that's something we can address later.

Is there any code that creates arrays of stuff to fit every coreid?

Reply:
> It's a bit user-unfriendly to have to write:
> `target create $_TARGETNAME_1 riscv -chain-position $_CHIPNAME.cpu -coreid 1048577`

`0x100001`  looks friendlier than `1048577`

> Is there any code that creates arrays of stuff to fit every coreid?

Sorry, I don't quite understand the question, could you explain it in more detail

Reply:
Is there any code that relies on the fact that coreids are usually small? E.g. somewhere maybe code creates an array indexed by coreid. That is fine if coreid numbers are small, but if all of a sudden that code needs to deal with core IDs like 0x100001, then memory is allocated for 0x100000 unused elements, which is excessively wasteful.

Reply:
I found no such code in openocd. And when I config 1 core with coreid 0, Windows Task Manager show that openocd cost 2.7MB, when config 4 core with coreid 0/1/0x100000/0x100001, openocd cost 4.4MB.

Reply:
Actually, why try to squeeze the core id and debug module base into the `coreid` variable in the first place? :)

> Tim: ... target->coreid is globally unique in OpenOCD ...

I believe there is no such requirement in OpenOCD for target->coreid to be globally unique. 

In principle, the coreid only needs to be unique for given RISC-V debug module -- identified by combination of `dbgbase` and `JTAG tap name`. (And if the user does not obey this rule, than it is a problem of the user-written configuration. We may add some checks for it one day, not necessarily in this merge request.)

Reply:
> Tim: But I don't think we need to worry about hartids right now. We do need to worry about debug index and OpenOCD coreid no longer being equal.

As indicated in the previous comment, I think this is an issue. Core IDs can remain equal to Hart IDs.

I believe that what needs to be dealt with is the `get_dm()` function that allocates new debug module instances. Currently, only JTAG chain position (`target->tap->abs_chain_position`) is taken into account. The value of `dbgbase` needs to be taken into account now, too.
https://github.com/riscv/riscv-openocd/blob/riscv/src/target/riscv/riscv-013.c#L232

Reply:
When I try to config all harts(from different DMs) into one smp, coreid shoule be globally unique because threadid comes from coreid
https://github.com/riscv/riscv-openocd/blob/21d21408aa36ccb458ccfaad871f7dd1bb1f731a/src/rtos/hwthread.c#L40-L43

But then I realized that config smp with harts from different DMs is not that easy, more things need to deal with, such as haltgroup. So this PR can ignore it.

I don't know yet if there are other places that require coreid to be globally unique.

Reply:
>  Core IDs can remain equal to Hart IDs.

This PR change ` info->index = target->coreid`  to ` info->index = target->coreid & (RISCV_MAX_HARTS - 1);`,
As coreid don't need to be globally unique, should I keep this change（like a protect, because up to 2^20 harts can be connected to a single DM）? or just drop the change, keep the original code  ` info->index = target->coreid` ?

Reply:
> I believe that what needs to be dealt with is the get_dm() function that allocates new debug module instances.

Thanks, that's really a problem, I'll handle it. There was no problem when debugging only one DM at a time, so I didn't notice it in time.

Reply:
>  As coreid don't need to be globally unique, should I keep this change（like a protect, because up to 2^20 harts can be connected to a single DM）? or just drop the change, keep the original code  info->index = target->coreid ?

I think it is OK to keep the current state: `info->index = target->coreid`

Reply:
ok, keep the current state make this PR simpler